### ISSUE TITLE
Not working with Logitech Options/Mouse

### ISSUE THREAD

**Issue by tmicky** [ther0n/UnnaturalScrollWheels#4]
I have the Logitech Wireless Trackball M570, and unfortunately this app isn't fixing the scroll direction. I'm not sure if the problem is because Logitech already controls mouse settings through its Options/device manager app (and it's _so_ frustrating that this mouse doesn't include the option to reverse scroll inside the Options app). I don't know anything about coding so I don't know how to figure out if the Options app is taking priority over your app or something, but I just wanted to let you know about a potential limitation. (Regardless of this issue your app looks _so_ cool and helpful, thanks so much for making it!!)

**Comment by ther0n** [comment]
Thanks for the report! I've never used a trackball mouse before (which also means I can't test it) so I'm not really sure how they work. It looks like you move the cursor around with the trackball, but there's also the regular two mouse buttons and scroll wheel. Is that correct?

And the part that isn't working is the normal scroll wheel between the two buttons? Or are you wanting to invert the direction the mouse moves when you use the trackball?

The Logitech Options app likely does interfere with the UnnaturalScrollWheels as well, espescially if it works in a similar way. UnnaturalScrollWheels intercepts scroll "events" (which occur each time you scroll on a mouse or trackpad). When the app recieves a scroll event it:

1. Determines if the event [is "continuous"](https://developer.apple.com/documentation/coregraphics/cgeventfield/scrollwheeleventiscontinuous), trackpads produce pixel-based/continuous scroll events and most mice produce line-based/non-continous scroll events.

2. If the event is continuous and, depending on settings, inverts the event and/or sets the lines scrolled of the event

3. Passes the modified event down to the system (or the next application that intercepts the event, possibly Logitech Options)

Logitech Options may be recieving the event then modifies the event *again* which could override or undo the changes UnnaturalScrollWheels made. It's also possible Logitech Options gets the event first and UnnaturalScrollWheels undoes the changes Logitech Options made.

It's also possible your mouse is simply sending pixel-based scroll events leading to UnnaturalScrollWheels not being aware it's a scroll wheel. There might be a few other ways it could differentiate between scroll wheels and trackpads, but I want to make sure I'm understanding the issue correctly first and ensure it not another application interfering/overriding.

**Comment by tmicky** [comment]
Yeah, the cursor is moved by rolling the trackball with a finger instead of from the movement of the entire mouse, but otherwise it has the same left/right click and scroll wheel. The part that isn't working is the regular scroll wheel, which scrolls the same way regardless of whether UnnaturalScrollWheels is turned on or not.

(Also I do want to be clear that I didn't bring this up to necessary get a fix/solution - I would not be surprised at all if the Logitech application is intercepting the event since it's scroll seems to be a little buggy anyway. I think the UnnaturalScrollWheels application is fantastic! And this is such a rare occurrence that would require the hardware to even figure out, which definitely wouldn't be worth it. I just wanted to mention it so you had a heads up of the potential limitation in this kind of rare instance.)

**Comment by ther0n** [comment]
While I don't have that particular mouse, I do have other Logitech mice and one of them was supported by Logitech Options. I set up both UnnaturalScrollWheels, and Logitech Options in a macOS virtual machine for testing and confirmed that Logitech Options is almost certainly the reason it isn't working for you. For one reason or another Logitech Options modifies scroll event's to be "continuous" which as I described earlier will cause UnnaturalScrollWheels to believe the event came from a trackpad.

I did a little bit of investigation and looked for other differences in between events that come from scroll wheels vs trackpads and found a few undocumented values that seem to only be non-zero for only trackpads. I used that to add a new "Alternate detection method" option to preferences to use these other values rather than checking whether the event is "continuous". It appears to work with Logitech Options in my testing in a virtual machine so hopefully, it will also work for you!

Install the attached version of the application, open preferences, check "Alternate detection method" and let me know if it works!

[UnnaturalScrollWheels.zip](https://github.com/ther0n/UnnaturalScrollWheels/files/5008629/UnnaturalScrollWheels.zip)
